### DISCUSSION TITLE
Create nick database entry in external program.

### DISCUSSION THREAD

**Discussion by abcbarryn** [anope/anope#441] [Support] [answered]
**Description**
I would like to be able to register a nick and set the password in an external program by adding a row to the MySQL DB table. What is the easiest way the duplicate Anope's password hashing algorithm?

**Why this would be useful**
I could have users register nicks by filling out a user-friendly web registration form.

**Comment by TehPeGaSuS** [comment]
Unless you really need to use an external program, Anope has a built-in webcpanel that allows you to register nicknames and manage accounts, channels, etc.

On Anope 2.0 you need to build webcpanel with `./extras` on the source folder. With Anope 2.1, it is built by default.

You could take a look at it and see if it fits your needs.

**Comment by abcbarryn** [comment]
I do eventually really need to use an external program because I want to link the user account to info in another database, but I will look at the webpanel and see if it can at least be used as a stop gap.

**Comment by abcbarryn** [comment]
Hmmm, I don't see webpanel listed in extras. I am running Anope 2.0.16...
```
$ ./extras
CMakeFiles                = wrong hardlink target (65025:2124835)
m_ldap.cpp                = disabled
m_ldap_authentication.cpp = disabled
m_ldap_oper.cpp           = disabled
m_mysql.cpp               = enabled
m_regex_pcre.cpp          = disabled
m_regex_pcre2.cpp         = disabled
m_regex_posix.cpp         = disabled
m_regex_tre.cpp           = disabled
m_sql_authentication.cpp  = disabled
m_sql_log.cpp             = disabled
m_sql_oper.cpp            = disabled
m_sqlite.cpp              = disabled
m_ssl_gnutls.cpp          = enabled
m_ssl_openssl.cpp         = disabled
stats                     = disabled

Please enter the name of the module or type 'q' to quit.: q
```

**Comment by TehPeGaSuS** [comment]
Right.
I think it was made to build by default some versions ago (I've been using 2.1 since when it came out, almost).
You'll probably only need to enable [httpd](https://github.com/anope/anope/blob/2.1/data/modules.example.conf#L201-L233), and [webcapnel](https://github.com/anope/anope/blob/2.1/data/modules.example.conf#L762-L786)

**Comment by Techman** [comment]
If you are moving towards a centralized account system that will span multiple services, it may be worth considering LDAP in the future.

**Comment by abcbarryn** [comment]
I got the Webcpanel working, although I am having isseues getting it to use TLS. It doesn't really fit what I need also. I could write something but I would need to be able to have it hash the password. Anope doesn't seem to be using a simple sha256 hash with a salt.

**Comment by Techman** [comment]
The way the password strings look depends on the choice of the primary "encryption" module. You can see how the format works by looking at the "OnEncrypt" function in the module file for whatever method you have chosen.

**Comment by abcbarryn** [comment]
I am looking at sha256.
I found this in the source code:
password format:  <hashmethod>:<password_b64>:<iv_b64>
I am guessing that the iv_b64 is some sort of salt and is base 64 encoded, but I haven't been able to replicate the hashing yet. A function like UnrealIRC's mkpasswd would be nice to have. If Anope would support UnrealIRC's Argon2 hash, I could use that but from what I have seen so far it does not. :(

**Comment by abcbarryn** [comment]
Looking at the hashed password, that looks like hex encoding, not base 64...?

**Comment by abcbarryn** [comment]
If in the config I have something like this:
```
module { name = "enc_sha256" }
module { name = "enc_none" }
```
With sha256 listed first as shown, the configuration documentation says:
 * The first encryption module loaded is the primary encryption module. All new
passwords are
 * encrypted by this module. Old passwords stored in another encryption method a
re
 * automatically re-encrypted by the primary encryption module on next identify.
So then if I use the above config, does that mean I could add the passwords in plain text and they would be automatically re-encrypted the first time the user logged in?

**Comment by sadiepowell** [comment]
If you want to authenticate against an external SQL database you should probably use the sql_authentication module rather than messing around with Anope's database.

**Comment by abcbarryn** [comment]
But Anope's NickServ already has a database, why not improve the registration process?

**Comment by sadiepowell** [comment]
I'm not sure what benefit it has? Anope has options for authenticating an external service against Anope (xmlrpc) and using an external service for authenticating against Anope (sql_authentication). It doesn't seem like a widely useful feature to maintain two databases.

**Comment by abcbarryn** [comment]
Maintaining two databases is exactly what I am trying to avoid. The NickServ database is required in order to have registered nicks, why setup another database just to authenticate nicks when the NickServ already has a database that does that? Plus the NickServ is (or at least was) clunky and required the users to confirm their email by typing text commands to IRC, I fised that so that now users get an email with a link that they can click to conform their email. All I need is better documentation on the sha256 password hash method and I could easily create a program to register and maintain nicks in an external program. This would allow integration with other systems that require users to register, like blogs, company logins, etc...

**Comment by abcbarryn** [comment]
The Web CPanel is good, but it's not easily extendable and the handling of email confirmations leaves much to be desired.

This is what a confirmation email for my system looks like now because of an external program I wrote...

```
From [services@amobiledevice.com]

Hi,

You have requested to register the nickname nick on AMobile IRC.
Please go to https://www.amobiledevice.com/confirm.php?name=nick&code=codehere
or type " /msg NickServ CONFIRM codehere " to complete the registration

If you don't know why this mail was sent to you, please ignore it silently.

AMobile IRC administrators.
```

As opposed to this:
```
From [services@amobiledevice.com]

Hi,

You have requested to register the nickname nick on AMobile IRC.
Please type " /msg NickServ CONFIRM codehere " to complete the registration

If you don't know why this mail was sent to you, please ignore it silently.

AMobile IRC administrators.
```

**Comment by sadiepowell** [comment]
It sounds like what you want is the xmlrpc module? You can create and authenticate a user using that.

https://github.com/anope/anope/tree/2.0/docs/XMLRPC

**Comment by abcbarryn** [comment]
Does anyone have a simple complete example program that would register a user, preferably written in PHP using XMLRPC? Just the basics, but something that I could run?

**Comment by sadiepowell** [comment]
There's one here: https://github.com/anope/anope/blob/2.1/docs/XMLRPC/xmlrpc.php#L61

**Comment by abcbarryn** [comment]
That has functions, but no example of how to use the functions. Also, I am guessing I would need to include that file using an includude statement from my actual code?

**Reply by sadiepowell** [reply]
I linked you to a line that shows you how to to call the function that registers a nickname. You need to edit the files to be usable on your own server.

**Reply by abcbarryn** [reply]
How do you pass the host parameter? Do you need to set the port? I am guessing you do something like:

```
<?php
include 'xmlrpc.php';
~ some code ~
$anope->command("ChanServ", "Adam", "REGISTER #adam");
?>
```
Also if your registering a nick wouldn't it be `$anope->command("NickServ", "Adam", "REGISTER password user@email.com");`?

**Comment by abcbarryn** [comment, marked as answer]
Ok, first off, I found that the xmlrpc extension is no longer bundled with PHP 8.x. You have to download it and build it separately. It is available here: https://github.com/php/pecl-networking-xmlrpc
To build it you need to run phpize in the folder where you extracted it, then run configure, then make, then make install. After that you will need to go to /etc/php8/conf.d and create a file named xmlrpc.ini containing the line:
extension=xmlrpc.so

Now your PHP has the XMLRPC extension loaded, copy [https://github.com/anope/anope/blob/2.1/docs/XMLRPC/xmlrpc.php](https://github.com/anope/anope/blob/2.1/docs/XMLRPC/xmlrpc.php) and edit the last line to contain the URL of your Anope web server instance. Make sure m_httpd is enabled and configured in your Anope modules.conf file. Also make sure m_xmlrpc and m_xmlrpc_main are uncommented and enabled. Now you can created a php file that includes xmlrpc.php and runs a command like this:
```
<?php
include 'xmlrpc.php';
$anope->command("NickServ", "testuser", "REGISTER password user@email.com");
?>
```
This example crates a nick name of "testuser" with a password of "password" and an email of "user@email.com".
